### ISSUE TITLE
Automatic and decoupled releases

### ISSUE THREAD

**Issue by ElectricMaxxx** [symfony-cmf/symfony-cmf#263]
I currently work on our dev-kit to get more controll over our releases. My main reasons are decoupling of the releases of the packages and simplify them at all. To get out faster in smaler steps. So what do i have in mind?

#### Independent releases
We talk about releasing each package in an independent manner, but do not do. We wait until each of the packages is ready and shout out then. This should be changed. Each package should release a new minor after mergin a complete feature into master. Each release on a package should trigger a release on sandbox also (see automation), doing so the sandbox should be fixed on explicit minor versions of our eco system and each single new release of i.e. seo-bundle would increase the version of seo-bundle in sandbox and of its own version. So the counter of sandbox versions would increase as the sum of all package's releases. 
This would have one outcome: each merge on master (or we do it like sonata and go on 1.x and 2.x to have master for major changes) should end in a green build on the packages, so the package is ready to ship, and on green build within the sandbox, so that the sandbox would be ready to ship. We would skip RC on minor releases and work with bug fixes instead. 
Why skipping RC? We currently have that less responses on our RC's that i would skip it and if something breaks, the user would hopefully report it, and would be able to switch back, as we would have more versions.
One other outcome would be: we should couple the docs inside the packages itself. Those common doc pages, which do not belong to any package at all, can still stay in a docs repository , which would be responseable to aggregate the complete docs. Having the docs in repo, would mean, that one Acceptence Criteria would be to fix or change docs. No more differences between docs and versions.

# Release automation
At the end the releasing should be done in a automatic manner, like we used to do it day by day in our jobs. So as i said, each package PR should be merged, when:

- docs are updated and valid
- own build is green
- integration build in sandbox is green (directly use the feature branch of the package itself)

having that, a merge on master would mean, a release on package, which should be done through CI and and a PR on sandbox to increase the package's version. We can add the new stuff of our feature then here. A green build and merge on master would end up in a new release of sandbox also.

That sounds like some work, but at the end we should care about merging features in our packages only. I would even add a automatic twitter notice on each release. And i really whant to have that automatic realeases to get back on track and get back some velocity and controll.

**Comment by dbu** [comment]
> independant releases & RC & sandbox

i agree with this. we have been way too reluctant to tag releases in the past, leading to features being ready in master but not used. we should not be excessive with version numbers, but much more liberal. not coupling minor versions between the different cmf bundles will make it easier to do that. and yeah lets not bother with RC that nobody looks at.

the sandbox basically does not need minor versions - we can do them or not, i don't mind. its a demo showcase of our stuff and not meant for people to depend on that git repository, or directly sync to it. we should composer update whenever one of the cmf components is tagged (be it a minor or a patch release) and then tag the sandbox (with a minor or patch - patch makes more sense for this to me) so that composer installs this latest version when asked to install a stable release.

> doc inside package repo

the problem i see with that is that our documentation is integrated on https://symfony.com/doc/ . also, documenting each component in isolation will lead to the overall documentation going out of sync. i'd prefer keeping it as is, with the `.. versionadded::` fields for new features.

> master vs 2.x etc branch

i would opt to only have the master branch and if a bundle needs a new major version, we can prepare that in a 3.x (or whatever the number is going to be) branch. i don't want to maintain several branches of stable and future version of the bundles, its a lot of effort and we have very limited resources.

> automated releases

i love the idea of more integration there, updating the sandbox and those things. i doubt that automatically releasing after each merge to master is a good idea. that is fine for custom applications where there is only one version existing. for an open source component, versions do have a bit of cost, and we need to distinguish between patches, minor and occasional major versions. if we can automate the process thats cool - but i'd prefer to trigger the process manually and not automatically.

**Comment by ElectricMaxxx** [comment]
> master vs 2.x etc branch
I could live with that, but this would mean we would have to wait for a major release, and won't be able to fix or simply add something on the latest stable major. I do not want to have more then two branches to manage in each package, one for latest stable and one to prepare a next major. I thought this would be possible by having a master wich goes ahead next major and a i.e. 2.x for latest stable stuff

> version in sandbox
I personally would see the sandbox as our integration environment. Having that in mind i would run tests against it on a feature branch of a package. Releasing it or putting some numbers on it wouldn't matter at all

> docs in package
But what about aggregation of our docs in a single docs repo, which could be a kind of mono repo? At the end we do have separate sections of each component and each bundle.  And sonata guys did some awesome effort in their dev-kit to test and integrate docs, which i would use in our fork also.
